# AI-Powered Retail Intelligence & Demand Forecasting Platform

## Module 1: Data Understanding

### Objective

Understand the dataset structure, business meaning, and data quality before performing cleaning, analysis, and machine learning.

---

#  Why Data Understanding?

Data Understanding is the first practical step after collecting a dataset.

Before applying any machine learning algorithm, a Data Scientist must understand:

- What information is available?
- What each row represents.
- What each column represents.
- Which columns are numerical and categorical.
- Whether there are missing values.
- Whether the data contains duplicate records.
- Whether the data types are correct.

Without understanding the dataset, any analysis or prediction may produce incorrect business decisions.

---

#### Step 1: Import Required Libraries

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

# Graph settings
plt.style.use("ggplot")
%matplotlib inline

#### Step 2: Loading / Reading the data 

In [2]:
df = pd.read_csv("../data/raw/Global_Superstore_2016.csv", encoding="latin1")
print(df)

       Row ID                  Order ID  Order Date   Ship Date  \
0       40098  CA-2014-AB10015140-41954  2014-11-11  2014-11-13   
1       26341    IN-2014-JR162107-41675  2014-02-05  2014-02-07   
2       25330    IN-2014-CR127307-41929  2014-10-17  2014-10-18   
3       13524   ES-2014-KM1637548-41667  2014-01-28  2014-01-30   
4       47221   SG-2014-RH9495111-41948  2014-11-05  2014-11-06   
...       ...                       ...         ...         ...   
51285   29002   IN-2015-KE1642066-42174  2015-06-19  2015-06-19   
51286   34337  US-2014-ZD21925140-41765  2014-05-06  2014-05-10   
51287   31315  CA-2012-ZD21925140-41147  2012-08-26  2012-08-31   
51288    9596   MX-2013-RB1979518-41322  2013-02-17  2013-02-21   
51289    6147   MX-2013-MC1810093-41416  2013-05-22  2013-05-26   

            Ship Mode   Customer ID       Customer Name      Segment  \
0         First Class  AB-100151402       Aaron Bergman     Consumer   
1        Second Class     JR-162107       Justin Ri

### Why do we load the dataset?

The dataset is loaded into a Pandas DataFrame so that it can be analyzed using Python.

A DataFrame is a two-dimensional tabular data structure consisting of rows and columns.

It provides powerful functions for:

- Viewing the data
- Cleaning the data
- Filtering records
- Performing calculations
- Building Machine Learning models

---

#### Step 3: Printing the first 5 rows; Why do we use df.head()?

The "head()" function displays the first five rows of the dataset.

It provides a quick overview of the dataset without displaying all records.

Using this function, we can:

- Verify whether the dataset has loaded correctly.
- Understand the structure of the data.
- Observe the column names.
- Check the format of values.
- Identify any obvious data quality issues.

This is usually the first function executed by every Data Scientist after loading a dataset.

---

In [3]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Postal Code,City,...,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Shipping Cost,Order Priority
0,40098,CA-2014-AB10015140-41954,2014-11-11,2014-11-13,First Class,AB-100151402,Aaron Bergman,Consumer,73120.0,Oklahoma City,...,TEC-PH-5816,Technology,Phones,Samsung Convoy 3,221.980,2,0.0,62.1544,40.77,High
1,26341,IN-2014-JR162107-41675,2014-02-05,2014-02-07,Second Class,JR-162107,Justin Ritter,Corporate,NaN,Wollongong,...,FUR-CH-5379,Furniture,Chairs,"Novimex Executive Leather Armchair, Black",3709.395,9,0.1,-288.7650,923.63,Critical
2,25330,IN-2014-CR127307-41929,2014-10-17,2014-10-18,First Class,CR-127307,Craig Reiter,Consumer,NaN,Brisbane,...,TEC-PH-5356,Technology,Phones,"Nokia Smart Phone, with Caller ID",5175.171,9,0.1,919.9710,915.49,Medium
3,13524,ES-2014-KM1637548-41667,2014-01-28,2014-01-30,First Class,KM-1637548,Katherine Murray,Home Office,NaN,Berlin,...,TEC-PH-5267,Technology,Phones,"Motorola Smart Phone, Cordless",2892.510,5,0.1,-96.5400,910.16,Medium
4,47221,SG-2014-RH9495111-41948,2014-11-05,2014-11-06,Same Day,RH-9495111,Rick Hansen,Consumer,NaN,Dakar,...,TEC-CO-6011,Technology,Copiers,"Sharp Wireless Fax, High-Speed",2832.960,8,0.0,311.5200,903.04,Critical


#### Step 4: Shape of the dataset 

It will show the total number of rows and columns in the dataset. 

in the exact format of (rows, columns)

The "shape" attribute returns the dimensions of the dataset.

It provides two values:

- Number of rows
- Number of columns

Knowing the size of the dataset helps estimate the complexity of analysis and machine learning tasks.

Large datasets generally require more computation and memory compared to smaller datasets.

---

In [4]:
df.shape

(51290, 24)

#### Step 5: Knowing all the columns of the dataset 

The `columns` attribute displays the names of all columns present in the dataset.

This helps us understand:

- Available features
- Business attributes
- Possible target variables
- Features useful for prediction

Understanding column names is essential before performing cleaning or feature engineering.

---

In [5]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Postal Code', 'City',
       'State', 'Country', 'Region', 'Market', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Shipping Cost', 'Order Priority'],
      dtype='str')

#### Step 6: df.info()

The `info()` function provides a concise summary of the dataset.

It displays:

- Total number of rows
- Total number of columns
- Column names
- Number of non-null values
- Data types
- Memory usage

This function helps identify incorrect data types and missing values before data preprocessing.

---

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Row ID          51290 non-null  int64  
 1   Order ID        51290 non-null  str    
 2   Order Date      51290 non-null  str    
 3   Ship Date       51290 non-null  str    
 4   Ship Mode       51290 non-null  str    
 5   Customer ID     51290 non-null  str    
 6   Customer Name   51290 non-null  str    
 7   Segment         51290 non-null  str    
 8   Postal Code     9994 non-null   float64
 9   City            51290 non-null  str    
 10  State           51290 non-null  str    
 11  Country         51290 non-null  str    
 12  Region          51290 non-null  str    
 13  Market          51290 non-null  str    
 14  Product ID      51290 non-null  str    
 15  Category        51290 non-null  str    
 16  Sub-Category    51290 non-null  str    
 17  Product Name    51290 non-null  str    
 1

#### Step 7 : df.describe

The `describe()` function generates statistical summaries of numerical columns.

It calculates:

- Count
- Mean
- Standard Deviation
- Minimum Value
- Maximum Value
- Quartiles

These statistics help understand the distribution of numerical features and identify unusual values or outliers.

---

In [7]:
df.describe

<bound method NDFrame.describe of        Row ID                  Order ID  Order Date   Ship Date  \
0       40098  CA-2014-AB10015140-41954  2014-11-11  2014-11-13   
1       26341    IN-2014-JR162107-41675  2014-02-05  2014-02-07   
2       25330    IN-2014-CR127307-41929  2014-10-17  2014-10-18   
3       13524   ES-2014-KM1637548-41667  2014-01-28  2014-01-30   
4       47221   SG-2014-RH9495111-41948  2014-11-05  2014-11-06   
...       ...                       ...         ...         ...   
51285   29002   IN-2015-KE1642066-42174  2015-06-19  2015-06-19   
51286   34337  US-2014-ZD21925140-41765  2014-05-06  2014-05-10   
51287   31315  CA-2012-ZD21925140-41147  2012-08-26  2012-08-31   
51288    9596   MX-2013-RB1979518-41322  2013-02-17  2013-02-21   
51289    6147   MX-2013-MC1810093-41416  2013-05-22  2013-05-26   

            Ship Mode   Customer ID       Customer Name      Segment  \
0         First Class  AB-100151402       Aaron Bergman     Consumer   
1        Second C

#### Step 8 : Null / Missing Values

Missing values represent unavailable or incomplete information.

If missing values are not handled properly, they may reduce model performance and produce inaccurate results.

Therefore, checking missing values is an essential part of Data Cleaning.

---

In [8]:
df.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Postal Code       41296
City                  0
State                 0
Country               0
Region                0
Market                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
dtype: int64

#### Step 9 : Checking Duplicated values 

Duplicate records represent repeated observations in the dataset.

Duplicates may occur because of:

- Manual data entry errors
- System errors
- Data merging issues

Duplicate records can affect statistical analysis and machine learning performance.

Therefore, they should always be identified before analysis.

---

In [9]:
df.duplicated().sum()

np.int64(0)

#### Step 10 : Checking data types

Every column in a dataset has a specific data type.

Examples include:

- Integer
- Float
- Object (Text)
- Boolean
- Datetime

Correct data types improve memory efficiency and ensure that analysis and machine learning algorithms work correctly.

Incorrect data types should be converted during the Data Cleaning stage.

---

In [10]:
df.dtypes

Row ID              int64
Order ID              str
Order Date            str
Ship Date             str
Ship Mode             str
Customer ID           str
Customer Name         str
Segment               str
Postal Code       float64
City                  str
State                 str
Country               str
Region                str
Market                str
Product ID            str
Category              str
Sub-Category          str
Product Name          str
Sales             float64
Quantity            int64
Discount          float64
Profit            float64
Shipping Cost     float64
Order Priority        str
dtype: object

### Final Observation: 

#### Summary & Key Findings from Data Understanding:

- Dataset Dimensions: Dataset contains 51,290 rows (transactions) and 24 columns (features).

- Missing Values:

        1. Postal Code has 41,296 missing values (~80.5%) because non-US regions in the Global Superstore dataset do not track US ZIP codes.

        2. All other 23 columns have 0 null values (100% complete data).

- Duplicates: Found 0 duplicate rows, confirming transaction uniqueness.

- Data Type Discrepancies:

        1. Order Date and Ship Date are currently stored as object (string) instead of datetime64.

        2. Postal Code is stored as float64.

Key Numerical Features: Sales, Quantity, Discount, Profit, and Shipping Cost are already available in numerical format (float64/int64).

---